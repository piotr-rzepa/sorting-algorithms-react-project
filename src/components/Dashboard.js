/*
 * Komponent bezklasowy będący stroną główną aplikacji
 */

import React, { useState, useContext } from 'react';
import Card from './Card';
import ChartBar from './ChartBar';
import ChartPie from './ChartPie';
import AlgoList from './AlgoList';
import SettingsModal from './SettingsModal';
import { ReactComponent as EditSVG } from '../images/edit.svg';
import GroupSVG from '../images/group.svg';
import ClockSVG from '../images/clock.svg';
import DecreasingSVG from '../images/decreasing.svg';
import TrophySVG from '../images/trophy.svg';
import { useSelector } from 'react-redux';
import selectFastestSlowestAlgo from '../selectors/fastslowestSelector';
import numeral from 'numeral';
import { ThemeContext } from './ThemeContext';
import { styleMain, styleButtons } from '../styles/styles';

export default function Dashboard() {
	// Pobieranie motywu [jasny/ciemny]
	const darkTheme = useContext(ThemeContext);

	//Ustawianie motywu w zależności od kontekstu
	const currentStyle = darkTheme ? styleMain.darkStyle : styleMain.lightStyle;
	const numberOfAlgorithms = useSelector((state) => state.algorithms.length);
	const [fastest, slowest, average] = useSelector((state) =>
		selectFastestSlowestAlgo(state)
	);
	const [modalIsOpen, setIsOpen] = useState(false);

	return (
		<div className="Dashboard" style={currentStyle}>
			<SettingsModal
				isOpen={modalIsOpen}
				onRequestClose={() => setIsOpen(false)}
				contentLabel="Example Modal"
			/>
			<div className="overview">
				<div className="overview-title">
					<h1>Dashboard</h1>
					<p>Overview of sorting algorithms</p>
				</div>
				<button
					className="data-options-button"
					type="button"
					onClick={() => setIsOpen(() => true)}
					style={darkTheme ? styleButtons.darkStyle : styleButtons.lightStyle}
				>
					<EditSVG width="24px" height="24px" />
					Manage data options
				</button>
			</div>
			<div className="small-cards">
				<Card
					title={numberOfAlgorithms}
					subtitle={'Algorithms'}
					icon={GroupSVG}
					background={{ backgroundColor: 'rgba(113,59,219,0.05)' }}
				/>
				<Card
					title={`${numeral(average * 100).format('00:00:00')}ms`}
					subtitle={'Overall average sorting time'}
					icon={ClockSVG}
					background={{ backgroundColor: 'rgba(76,184,255,0.07)' }}
				/>
				<Card
					title={slowest}
					subtitle={'Worst average sorting time'}
					icon={DecreasingSVG}
					background={{ backgroundColor: 'rgba(255,76,97,0.05)' }}
				/>
				<Card
					title={fastest}
					subtitle={'Best average sorting time'}
					icon={TrophySVG}
					background={{ backgroundColor: 'rgba(255,184,0,0.07)' }}
				/>
			</div>
			<div className="charts">
				<ChartBar width={820} height={609} />
				<AlgoList />
				<ChartPie width={387} height={609} />
			</div>
			<p style={{ textAlign: 'right', padding: '2px 0' }}>
				Made by Chili Labs & Piotr
			</p>
		</div>
	);
}
