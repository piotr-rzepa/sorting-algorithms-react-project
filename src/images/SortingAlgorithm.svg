<svg enable-background="new 0 0 24 24" height="512" viewBox="0 0 24 24" width="512" xmlns="http://www.w3.org/2000/svg"><path d="m17.5 0h-15.75c-.965 0-1.75.785-1.75 1.75v1.926c0 .735.286 1.426.806 1.945l5.828 5.824c.232.232.366.555.366.884v7.921c0 .293.17.559.437.681.1.047.207.069.313.069.176 0 .35-.062.488-.181l2.801-2.401c.611-.522.961-1.284.961-2.088v-3.98c0-.337.139-.666.382-.9l6.027-5.815c.534-.515.84-1.237.84-1.979v-1.906c.001-.965-.784-1.75-1.749-1.75z"/><path d="m22.25 12h-6.5c-.965 0-1.75.785-1.75 1.75v8.5c0 .965.785 1.75 1.75 1.75h6.5c.965 0 1.75-.785 1.75-1.75v-8.5c0-.965-.785-1.75-1.75-1.75zm-2 8.5h-2.5c-.414 0-.75-.336-.75-.75s.336-.75.75-.75h2.5c.414 0 .75.336.75.75s-.336.75-.75.75zm0-3h-2.5c-.414 0-.75-.336-.75-.75s.336-.75.75-.75h2.5c.414 0 .75.336.75.75s-.336.75-.75.75z"/></svg>