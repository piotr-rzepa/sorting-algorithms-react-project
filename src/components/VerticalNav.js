/*
 * Komponent bezklasowy będacy menu pionowym
 * Pozwala przejść z menu głównego na stronę ...
 */

import React from 'react';
import { Link } from 'react-router-dom';
import { ReactComponent as DashboardSVG } from '../images/Dashboard.svg';
import { ReactComponent as SelectedAlgorithmSVG } from '../images/SortingAlgorithm.svg';
import { ReactComponent as SettingsSVG } from '../images/settings.svg';

function VerticalNav() {
	return (
		<div className="wrapper">
			<div className="verticalNav">
				<ul>
					<li>
						<Link to="/">
							<DashboardSVG width="24px" height="24px" className="svg-image" />
						</Link>
					</li>
					<li>
						<Link to="/algorithm/">
							<SelectedAlgorithmSVG width="24px" height="24px" className="svg-image" />
						</Link>
					</li>
					<li>
						<hr />
						<a href="#" className="settings">
							<SettingsSVG height="24px" width="24px" className="svg-image" />
						</a>
					</li>
				</ul>
			</div>
		</div>
	);
}

export default VerticalNav;
