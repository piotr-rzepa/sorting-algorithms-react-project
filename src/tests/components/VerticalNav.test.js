import React from 'react';
import { shallow } from 'enzyme';
import VerticalNav from '../../components/VerticalNav';
import { Link } from 'react-router-dom';
import { ReactComponent as DashboardSVG } from '../../images/Dashboard.svg';
import { ReactComponent as SelectedAlgorithmSVG } from '../../images/SortingAlgorithm.svg';
import { ReactComponent as SettingsSVG } from '../../images/settings.svg';

//* Sztuczna implementacja moduÅ‚u react-router-dom
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useLocation: () => ({ pathname: 'localhost:3000' })
}));

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useContext: jest.fn().mockImplementation(() => ({ darkTheme: true }))
}));

describe('<VerticalNav />', () => {
	it('Should render Horizontal navbar correctly', () => {
		const wrapper = shallow(<VerticalNav />);
		expect(wrapper).toMatchSnapshot();
	});

	it('Should render expected amount of children components', () => {
		const wrapper = shallow(<VerticalNav />);
		expect(wrapper.find('.wrapper')).toHaveLength(1);
		expect(wrapper.find('.verticalNav')).toHaveLength(1);
		expect(wrapper.find(DashboardSVG)).toHaveLength(1);
		expect(wrapper.find(SelectedAlgorithmSVG)).toHaveLength(1);
		expect(wrapper.find(SettingsSVG)).toHaveLength(1);
		expect(wrapper.find(Link)).toHaveLength(2);
	});

	it('Should handle toggle theme by button click', () => {
		const wrapper = shallow(<VerticalNav />);
		wrapper.find('.settings').simulate('click');
		expect(React.useContext).toHaveBeenCalledTimes(1);
	});

	it('Should have className', () => {
		jest.mock('react-router-dom', () => ({
			...jest.requireActual('react-router-dom'),
			useLocation: () => ({ pathname: 'localhost:3000' })
		}));
		const wrapper = shallow(<VerticalNav />);
		expect(wrapper.find(Link).at(0).hasClass('current')).toEqual(false);
	});

	it('Should not have className', () => {
		jest.mock('react-router-dom', () => ({
			...jest.requireActual('react-router-dom'),
			useLocation: () => ({ pathname: '/algorithm/' })
		}));
		const wrapper = shallow(<VerticalNav />);
		expect(wrapper.find(Link).at(1).hasClass('current')).toEqual(false);
	});
});
