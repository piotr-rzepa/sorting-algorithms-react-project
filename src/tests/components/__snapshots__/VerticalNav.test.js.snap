// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerticalNav /> Should render Horizontal navbar correctly 1`] = `
<div
  className="wrapper"
>
  <div
    className="verticalNav"
  >
    <ul>
      <li>
        <Link
          className=""
          to="/"
        >
          <ForwardRef(SvgDashboard)
            className="svg-image"
            height="24px"
            width="24px"
          />
        </Link>
      </li>
      <li>
        <Link
          className="disabled"
          to="#"
        >
          <ForwardRef(SvgSortingAlgorithm)
            className="svg-image"
            height="24px"
            width="24px"
          />
        </Link>
      </li>
      <li>
        <hr />
        <button
          className="settings"
          type="button"
        >
          <ForwardRef(SvgSettings)
            className="svg-image"
            height="24px"
            width="24px"
          />
        </button>
      </li>
    </ul>
  </div>
</div>
`;
